### ISSUE TITLE
Schema diff keeps thinking custom mapping types have changed

### ISSUE THREAD

**Issue by vicdelfant** [doctrine/dbal#2596]
I'm working on a relatively large project with a lot of custom mapping types (primarily value objects with UUIDs such as `OrderId`, `InvoiceId` etc.). For some reason, `schema:update` insists on re-creating all columns that contain such a custom type:

```bash
$ bin/console doctrine:schema:update --dump-sql
ALTER TABLE orders_order CHANGE order_id order_id CHAR(36) NOT NULL COMMENT '(DC2Type:orders.order_id)';
# ...

$ bin/console doctrine:schema:update --force
Updating database schema...
Database schema updated successfully! "33" queries were executed

$ bin/console doctrine:schema:update --dump-sql
ALTER TABLE orders_order CHANGE order_id order_id CHAR(36) NOT NULL COMMENT '(DC2Type:orders.order_id)';
# ...
```

Mapping types are properly registered, `requiresSQLCommentHint` is set to return `true` for all types, the `DC2Type` comment is correctly being added to the field definition (confirmed in the MySQL tables), storing and retrieving data works perfectly... everything but the schema comparison.

The types are registered through a dedicated service, but in a nutshell the process is as follows:
```php
foreach ($types as $name => $class) {
    Type::addType($name, $class);

    $platform->registerDoctrineTypeMapping($name, $name);

    // Added explicit call to markDoctrineTypeCommented(), makes no difference
    $platform->markDoctrineTypeCommented($name);
}
```

`AbstractPlatform::$doctrineTypeComments` contains all types so they're being processed just fine. Re-creating the entire database doesn't help either, and this obviously also affects migrations and `schema:validate`.

What am I missing here?

I came across the following similar issues but none of them solved my issue:
* https://github.com/sonata-project/SonataUserBundle/issues/103#issuecomment-10083314
* https://github.com/doctrine/dbal/issues/1614
* https://blog.vandenbrand.org/2015/06/25/creating-a-custom-doctrine-dbal-type-the-right-way/

**Comment by Ocramius** [comment]
Possibly strictly related to https://github.com/doctrine/dbal/issues/2594

**Comment by vicdelfant** [comment]
I'll try to find some time to write a test case so you can reproduce the issue more easily.

**Comment by deeky666** [comment]
@vicdelfant do you derive that particular custom type from a built-in concrete Doctrine type like `StringType`? We have some special comparisons for some types which could affect derived custom types. Maybe you can provide the code of that custom type? That would help a lot. Thanks.

**Comment by vicdelfant** [comment]
It's not derived from a built-in type, but the type class does extend quite a few of our own types before extending `Doctrine\DBAL\Types\Type`. The simplified `OrderId` and `OrderIdType` classes can be found at https://github.com/vicdelfant/doctrine-issue-2596, so it's basically comparable to https://github.com/ramsey/uuid-doctrine.

I our actual project, however, the inheritance chain is as follows:

1. `ProjectName\OrderIdType`
2. `ProjectName\AbstractGuidType`
3. `ProjectName\AbstractStringType`
4. `ProjectName\AbstractScalarType`
5. `Doctrine\DBAL\Types\Type`

As you can see, we love inheritance 😉 If you suspect inheritance to be the culprit then let me know and I'll update the test repo so you can get a complete picture.

**Comment by deeky666** [comment]
@vicdelfant I think your problem is [this line](). It won't match `orders.order_id` and thus not recognize that there is a custom type declaration in the comment. A workaround might be to choose another name for now. Not sure if we can widen the list of allowed characters. @Ocramius what do you think?

**Comment by vicdelfant** [comment]
Ah! Adding a `\.` to that regex indeed solves the issue for all `OrderIdType`-like types.

It appears that there's a second issue with a custom type inheriting from `DateTimeTzType` though; the comment field does seem to get through to the database schema, even with `requiresSQLCommentHint` set to `true`. I guess that's what you meant in your earlier comment? The custom type doesn't do anything apart from casting it to a `Carbon` instance.

**Comment by deeky666** [comment]
@vicdelfant I don't quite understand the second issue. Isn't it expected behaviour that a column comment is created in the database if `requiresSQLCommentHint` is set to `true`?

**Comment by vicdelfant** [comment]
Yes, exactly, that's the weird thing... there's no DC2Type-comment, even with `requiresSQLCommentHint` set to `true`.

**Comment by deeky666** [comment]
Maybe related to #2594?

**Comment by vicdelfant** [comment]
I'll check against `master` and report back, thanks!

**Comment by Sharom** [comment]
I have the same problem when type is a child of `StringType`.

**Type code example:**
```php
final class PhoneNumberType extends StringType
{
    const NAME = 'phone_number';

    public function getName()
    {
        return self::NAME;
    }

    public function getSQLDeclaration(array $fieldDeclaration, AbstractPlatform $platform)
    {
        return parent::getSQLDeclaration(
            array_merge(
                $fieldDeclaration,
                [
                    'length' => 16,
                    'fixed' => false
                ]
            ),
            $platform
        );
    }

    public function convertToDatabaseValue($value, AbstractPlatform $platform)
    {
        if ($value === null) {
            return $value;
        }

        if (!$value instanceof PhoneNumber) {
            throw new \InvalidArgumentException(sprintf(
                'Value must be of type "%s" but "%s" was given.',
                PhoneNumber::class,
                is_object($value) ? get_class($value) : gettype($value)
            ));
        }

        return $value->getRawValue();
    }

    public function convertToPHPValue($value, AbstractPlatform $platform)
    {
        if ($value === null) {
            return $value;
        }

        return new PhoneNumber($value);
    }

    public function requiresSQLCommentHint(AbstractPlatform $platform)
    {
        return true;
    }
}
```

**Comment by ragboyjr** [comment]
@NaGeL182 or anyone that is experiencing an issue similar to this with doctrine and symfony:

I was experiencing this issue as well for a custom type that extended the StringType. I didn't update the `getName` method of my custom type and it was writing the type as string into the sql comment which was what was causing the issue. Making sure my `getName` represented the name of my custom type (`channel_enum` in my case), doctrine's mapping tool worked perfectly.

**Comment by NaGeL182** [comment]
@ragboyjr thanks for the info but I still encounter it even if my `getName` method is updated!

```php
    const MONEYTYPE = 'money';
    //...

    public function getName()
    {
        return self::MONEYTYPE;
    }
```

and still encounter this error!

**Comment by NaGeL182** [comment]
@vicdelfant 
I just set this:
```php
    public function getName()
    {
        return self::MONEYTYPE;
    }

    public function requiresSQLCommentHint(AbstractPlatform $platform)
    {
        return true;
    }

    public function canRequireSQLConversion()
    {
        return true;
    }
```
and it still does it.

**Comment by NaGeL182** [comment]
hey, this seems to be fixed in `doctrine/orm` `v2.6.2`!
I don't get those diffs anymore! dunno what fixed it tho because the dbal is still at `2.8.0`...

**Comment by KartaviK** [comment]
> @vicdelfant
> I just set this:
> 
> ```
>     public function getName()
>     {
>         return self::MONEYTYPE;
>     }
> 
>     public function requiresSQLCommentHint(AbstractPlatform $platform)
>     {
>         return true;
>     }
> 
>     public function canRequireSQLConversion()
>     {
>         return true;
>     }
> ```
> 
> and it still does it.

I made value object as dbal type, and every field in mapping file represents as different every time. This overrides helps me to prevent creating same migrations for diff when mapping files same every time. thx
